src/js/angular/clustermanagement/controllers.js

```diff
@@ -12,765 +12,184 @@ const modules = [
+        $scope.errors.splice(0);
+        if (status === 412) {
+            $scope.preconditionErrors = Object.keys(data).map((key) => `${key} - ${data[key]}`);
+        } else if (status === 400) {
+            $scope.errors.push(...data);
+        } else if (status === 409) {
+            $scope.errors.push(data);
         }
-    };
-
-    $scope.getNodes();
-
-    $scope.setMode = function (mode) {
-        $scope.attributeChange = true;
-        $scope.setMasterAttribute($scope.selectedNode, 'Mode', mode);
-    };
+    }
 
-    $scope.setReadonly = function (readonly) {
-        $scope.attributeChange = true;
-        $scope.selectedNode.readonly = readonly;
-        $scope.setMasterAttribute($scope.selectedNode, 'ReadOnly', readonly);
-    };
+    function validateClusterSettings() {
+        const isFormValid = !$scope.clusterConfigurationForm.$invalid;
+        const nodesListValid = $scope.clusterConfiguration.nodes && $scope.clusterConfiguration.nodes.length >= 2;
+        return isFormValid && nodesListValid;
+    }
 
-    let timerCounter = 0;
-    const timer = $interval(function () {
-        // check for update only if not in the middle of set attribute operation
-        if (!$scope.attributeChange && !$scope.updating) {
-            timerCounter++;
-            // every third iteration also checks for added/removed nodes
-            $scope.updateClusterIfNeeded(timerCounter % 3 === 0);
+    $scope.setLoader = function (loader, message) {
+        $timeout.cancel($scope.loaderTimeout);
+        if (loader) {
+            $scope.loaderTimeout = $timeout(function () {
+                $scope.loader = loader;
+                $scope.loaderMessage = message;
+            }, 300);
+        } else {
+            $scope.loader = false;
         }
-    }, 2000);
-
-    $scope.$on("$destroy", function () {
-        $interval.cancel(timer);
-    });
-
-    $scope.attributeChange = false;
-    $scope.updating = false;
-
-    // track window resizing
-    const w = angular.element($window);
-    const resize = function () {
-        $scope.resize();
     };
-    w.bind('resize', resize);
-    $scope.$on('$destroy', function () {
-        w.unbind('resize', resize);
-    });
 
-    // track collapsing the nav bar
-    $scope.$on('onToggleNavWidth', function () {
-        $scope.resize();
-    });
-}
-
-CloneRepositoryCtrl.$inject = ['$scope', '$modalInstance', 'clone'];
-
-function CloneRepositoryCtrl($scope, $modalInstance, clone) {
-
-    $scope.clone = angular.copy(clone);
-    $scope.ok = function () {
-        const patt = new RegExp('^[a-zA-Z0-9-_]+$');
-        $scope.isInvalidRepoName = !patt.test($scope.clone.repositoryID);
-        if (!$scope.isInvalidRepoName) {
-            $modalInstance.close($scope.clone);
+    $scope.addNodeToList = function (nodeRpcAddress) {
+        if (!nodeRpcAddress || $scope.clusterConfiguration.nodes.includes(nodeRpcAddress)) {
```

Comment:
Shouldn't there be some notification in this case?

Reply:
There is a visual notification under the box that is shown when node is already in the list - `class="nodeAlreadyAdded". This is just a check in the event callback